### PR TITLE
feat: Design System Setup

### PR DESCRIPTION
**Pull request by agg23**
### User facing changelog
- Introduces the basics of the new Cypress design system. Not user facing yet, but we will start seeing changes across products to bring more consistency to our UI.
- New search component in Component Testing

### Additional details
Cypress's UI components are not as strictly conformant to a particular style or set of colors as we would like. This PR sets up the basis of a shared design system for all of Cypress, starting with the runners (Component Testing in particular).

### Principles
- Pixel perfect - There should be perfect, or imperceptibly unperfect, sizing and positioning of all elements. Fractional pixel sizes should be avoided at all costs, and text and images should always be crisp and appropriately sized
- Tokenized - Everything should be tokenized and reusable. No pixels or static CSS attributes unless absolutely necessary
- Semantic naming - Consistent names that give developers a sense of what the token means/is intended for.
- Composable - Whenever possible (particularly in the Core library), components and styles should gracefully cascade, allowing existing functionality and design to be reused
- Accessible - All components should enforce ARIA tags to indicate the purpose of elements, and keyboard navigation should be a first class citizen

#### CSS

The basis of the system is a set of variables in SASS that are used to generate various classes, variables, and mixins. These are split between the core design tenants: color, spacing, surfaces, and typography. All positioning is based off of `rem` units, with our base font size set at `16px = 1rem`. 16px was chosen because it allows for pixel perfect scaling with common ratios, and it has the added benefit that most browsers agree with this by setting their default `font-size: 16px`. We use "t-shirt" sizes (`xs, s, ms, m, ml, l, xl, 2xl, 3xl, 4xl`) as our semantic naming for most size related tokens.

Of note is the `css/derived` folder, which contains SASS files that exist purely to export values to JavaScript. These are mostly just used for tests/Storybook stories (which are generated based on the SASS), but we also use them to generate TypeScript types and to populate some class names. This brings the benefit of deriving all token references throughout the JS codebase to a single SASS file. Changing the name of a variable (or even adding a new one) in the `typography.scss` core file will automatically generate the classes and mixins used throughout the system, and will update the TypeScript types to enable support in JS.

Core variables that may need to be user changeable (only colors at the time of writing) are entirely expressed using CSS variables. The SASS variables exposed for each color actually return `var(--color-name)` to prevent developers from accidentally including a color literal. In time, these variables could lead to future color developments, particularly a dark theme.

#### Typography

Of particular note is the way the system handles type and iconography layout. The styles are customized to offer near-pixel perfect alignment of icon top and baseline. While I am very happy with how this is implemented, it may produce problems for us in the future, so dev users should beware.

Screen Shot 2021-04-02 at 1 01 41 PM

#### Core

The system provides several primitives to make integrating the system styles and tokens simpler. These primitives are strongly typed and provide easy customization to overall size, color, and other common values. Additionally, functionality-related primitives such as `Input` and `Button` come out of the box with support for common functionality and ARIA integration. Where necessary, types will enforce inclusion of ARIA props, and dev warnings will be logged in other non-compliant situations. These components should also support keyboard navigation out of the box.

#### Components

While there are no true Cypress components exposed yet, as the needs of Cypress's products evolve, we will be publishing composite components that wrap styles and functionality for more complex scenarios. These will provide component-level consistency where possible and provide a base for sharing core business logic (such as how Cypress handles directories and specs) between experiences.

### How has the user experience changed?

Component testing has received a different search box, taken from the design system. No functionality changes are present.

Screen Shot 2021-04-02 at 11 53 45 AM

### CODE REVIEW

File: npm/design-system/src/components/searchInput/SearchInput.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import { useCallback, useState } from 'react'
+
+describe('SearchInput', () => {
+  const StatefulWrapper: React.FC<{onInput?: (input: string) => void}> = ({ onInput }) => {
+    const [value, setValue] = useState('')
+
+    const memoedOnInput = useCallback((input: string) => {
+      setValue(input)
+      onInput?.(input)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (lmiller1990) [inline npm/design-system/src/components/searchInput/SearchInput.spec.tsx:14]: How come you don't pass the deps to this hook?

File: npm/design-system/src/components/searchInput/SearchInput.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import * as React from 'react'
+
+import { createStory, createStorybookConfig } from '../../stories/util'
+
+import { SearchInput as SearchInputComponent } from './SearchInput'
+import { useState } from 'react'
```
Comment (lmiller1990) [inline npm/design-system/src/components/searchInput/SearchInput.stories.tsx:6]: Any reason import react on two different lines?

File: npm/design-system/src/components/searchInput/SearchInput.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import * as React from 'react'
+import { FormEvent, MutableRefObject, useCallback } from 'react'
```
Comment (lmiller1990) [inline npm/design-system/src/components/searchInput/SearchInput.tsx:2]: Can't we do `import React, { otherStuff } from 'react'`?

File: npm/design-system/src/components/searchInput/SearchInput.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import { IconInput, IconSettings } from '../../core/input/IconInput'
+import { CoreComponent } from '../../core/shared'
+import { TextSize } from '../../css'
+import { useCombinedRefs } from '../../hooks/useCombinedRefs'
+
+export interface SearchInputProps extends CoreComponent {
```
Comment (lmiller1990) [inline npm/design-system/src/components/searchInput/SearchInput.tsx:8]: Would it make sense for our component props to extend `React.HTMLAttributes<T>` so we can easily pass whatever applicable attributes?

File: npm/design-system/src/core/button/Button.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      <ButtonComponent aria-label="buttonPress" color='white' onPress={action('buttonPress')}>Simple button</ButtonComponent>
+      <LinkButton aria-label="anchorButtonPress" color='white' onPress={action('anchorButtonPress')}>Anchor button</LinkButton>
+    </PaddedBox>
+  </div>
+))
+
+export const ButtonSizes = createStory(() => (
+  <div>
+    <div style={{ width: 500 }}>
+      {Object.keys(typography).filter((key) => key !== 'type' && !key.startsWith('line-height') && !key.startsWith('text-mono')).map((key) => {
```
Comment (lmiller1990) [inline npm/design-system/src/core/button/Button.stories.tsx:38]: Nit: we could use `reduce` here to have a lot less loops

File: npm/design-system/src/css/derived/types.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable */
+import type colors from './jsColors.scss'
+import type { ClassNames as JsSpacing } from './jsSpacing.scss'
+import type { ClassNames as JsTypography } from './jsTypography.scss'
+import type {ClassNames as JsSurface } from './jsSurfaces.scss'
+
+type Digit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'
+type TwoDigit = `${Digit}${Digit}`
```
Comment (lmiller1990) [inline npm/design-system/src/css/derived/types.ts:8]: Ah, this is the new TS feature you wanted?

**Review by lmiller1990 — DISMISSED** [review]
Left some questions, the PR is quite big to review in great detail, but I think the setup seems reasonable.

File: npm/design-system/src/css/derived/types.ts (hunk 6 above)
Comment (agg23) [reply npm/design-system/src/css/derived/types.ts:8]: This is indeed the TS feature, among a few other things.

File: npm/design-system/src/components/searchInput/SearchInput.tsx (hunk 4 above)
Comment (agg23) [reply npm/design-system/src/components/searchInput/SearchInput.tsx:8]: Potentially, yes. For now, I'm trying to lock that down as much as possible (can't deviate from the system semantics if you can't send arbitrary props to the DOM), but I may change this.

File: npm/design-system/src/components/searchInput/SearchInput.tsx (hunk 3 above)
Comment (agg23) [reply npm/design-system/src/components/searchInput/SearchInput.tsx:2]: Yes. Bad habit.

File: npm/design-system/src/components/searchInput/SearchInput.spec.tsx (hunk 1 above)
Comment (agg23) [reply npm/design-system/src/components/searchInput/SearchInput.spec.tsx:14]: This is a test and `onInput` should not be changing

File: npm/design-system/src/core/button/Button.stories.tsx (hunk 5 above)
Comment (agg23) [reply npm/design-system/src/core/button/Button.stories.tsx:38]: I would be interested in seeing what you suggest. This is a story though, so I don't think we're very concerned with performance.

Comment (lmiller1990) [reply npm/design-system/src/core/button/Button.stories.tsx:38]: Completely untested, might have the condition backwards, but something like:

```ts
{typography.reduce((acc, key) => {
  if (key !== 'type' && !key.startsWith('line-height') && !key.startsWith('text-mono')) {
    return acc
  }

  const size = key.replace('text-', '')
 
  return acc.concat(
    return (
      <ButtonComponent
        key={key}
        size={size as TextSize}
        aria-label="buttonPress"
      >
         `Button ${size}`}
      </ButtonComponent>
    )
  )
},[])}
```

Basically whenever I see filter + map I assume they can be combined into reduce. That said, I normally do this for perf and I don't see any real problem either way.

File: npm/design-system/src/css/derived/types.ts (hunk 6 above)
Comment (lmiller1990) [reply npm/design-system/src/css/derived/types.ts:8]: I'm constantly amazed at how far TS has come (and how it keeps evolving).

**Comment by lmiller1990** [comment]
Attempting to use React 17 is giving us major problems. I downgraded to the design system to React 16.8. I think this is fine for now, we can upgrade everything to React 17 at the same time.

Also, because the design system uses Vite for the dev server, and there is known problems with Vite on CI (via yarn cypress run-ct) I removed it from CI for now. We can add it back in when we fix the Vite dev server cold start issue.

**Review by lmiller1990 — APPROVED** [review]

**Pull request merged**